Add UPE feature flag controller

harriswong opened the pull request:
# Changes proposed in this Pull Request:

Issue: 
Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1684

Description: 
Adding a new end point `wp-json/wc/v3/wc_stripe/upe_flag_toggle` to toggle UPE feature. This will be used in UI so that users can click a button to enable UPE.

This end point is only available when `_wcstripe_feature_upe` is enabled.


# Testing instructions
1. You can test the API end point by using Postman and the API key generated from your localhost per instructions here (https://docs.woocommerce.com/document/woocommerce-rest-api/) 
2. Make sure you have `_wcstripe_feature_upe` flag enabled.
3. Use postman to go do a `GET` request to `https://localhost/wp-json/wc/v3/wc_stripe/upe_flag_toggle`. **Note: https**
4. You should see a JSON response `{"is_upe_enabled": "no"}`
5. If you want, you can double check your database `select * from wp_options where option_name ='woocommerce_stripe_settings'`. You should be able to see the serialized string and you can find the `is_upe_enabled` value there.
6. Make a `POST` request to `https://localhost/wp-json/wc/v3/wc_stripe/upe_flag_toggle` with body 
```
{
    "is_upe_enabled": "yes"
}
```
7. You should see a `200` response `{"result":"success"}`. 
8. Use postman to go do a `GET` request again to `https://localhost/wp-json/wc/v3/wc_stripe/upe_flag_toggle`.
9. You should see a JSON response `{"is_upe_enabled": "yes"}`. The flag has been updated.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Add UPE feature flag controller

EvanDotPro commented:
This may be known, but I just want to clarify that there are two layers for actually enabling UPE. The feature flag and the stripe setting.

The feature flag simply makes the actual merchant-facing stripe setting to enable or disable UPE visible, presumably only until UPE is GA in a release. The feature flag toggling provided by the controller in this PR is handled via WCPay Dev Tools on WCPay. See details here: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1646#discussion_r688855150

Personally, I didn't see any need for a UI or URL to toggle the feature flag, as running a quick upsert on the db is simple enough, but I won't object if this makes it easier for folks.

harriswong commented:
Thanks for the early feedback @EvanDotPro, appreciate it!

The purpose of this end point is to allow user to toggle the UPE feature through the UI CTA component, something like this:
![image](https://user-images.githubusercontent.com/572862/129791707-78ddc7d9-0f38-421e-9b30-2e77e76fbd09.png)

By clicking the "Enable in your store", it will call this end point to update the flag.

Thanks for the link to [the PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1646), I notice there are existing flags there. The one we want to toggle here should be `upe_checkout_experience_enabled`.

EvanDotPro commented:
> The one we want to toggle here should be `upe_checkout_experience_enabled`.

Yep that is correct since this is for the merchant-facing toggle. Glad I checked, I suspected this was the case. :)

EvanDotPro commented:
**Note:** The `feature/upe` branch has now been merged into `develop`. You can go ahead and rebase this branch against the latest `develop`, change the target branch of this PR, and force push the rebased branch.

harriswong commented:
Thanks for the update!

Commit:
Use existing feature flag upe_checkout_experience_enabled

Save this feature flag into the entire settings `woocommerce_stripe_settings`.
This also triggers pre_update_option_woocommerce_stripe_settings hook.

Commit:
Add unit test for UPE get/set flag in controller

Commit:
Compatibility with wp 4.7

has_params is not introduced until 5.3.0

Commit:
Fix Yoda Condition checks, I must.

Commit:
Use constant for flag name

frosso commented on includes/admin/class-wc-rest-upe-flag-toggle-controller.php:
```diff
@@ -0,0 +1,105 @@
+
+		$settings = get_option( 'woocommerce_stripe_settings' );
+
+		if ( $is_upe_enabled ) {
```
It looks like the `get_flag` returns a string `'yes'`/`'no``. But this method accepts a boolean to determine if the same flag should be enabled/disabled.

What do you think about ensuring that both methods accept/return consistent values (either both accept booleans or both accept `'yes'`/`'no'` strings)?

harriswong replied:
Updating API payload to `{"is_upe_enabled": "yes"}` instead of boolean.

frosso commented on woocommerce-gateway-stripe.php:
```diff
@@ -544,19 +544,31 @@ public function add_emails( $email_classes ) {
+				$controllers = [
+					'WC_Stripe_Connect_REST_Oauth_Init_Controller' => [ $this->connect, $this->api ],
+					'WC_Stripe_Connect_REST_Oauth_Connect_Controller' => [ $this->connect, $this->api ],
+					'WC_REST_UPE_Flag_Toggle_Controller' => null,
```
Should this register only when the `_wcstripe_feature_upe` flag (which controls the UPE preview) is enabled?

harriswong replied:
I think it makes sense to wrap around `_wcstripe_feature_upe` flag because there is no need to load this controller otherwise.

Commit:
Use 'yes' 'no' instead of true false for API

Commit:
Do not init WC_REST_UPE_Flag_Toggle_Controller if _wcstripe_feature_upe is off

frosso reviewed (approved):
Thanks for all the changes, LGTM 👍
